src/debugger/mock-ui/session.js

```diff
@@ -0,0 +1,239 @@
+  // command to start Prepack with
+  _prepackCommand: string;
+
+  _startAdapter() {
+    let adapterArgs = [this._adapterPath, "--prepack", this._prepackCommand];
+    this._adapterProcess = child_process.spawn("node", adapterArgs);
+    this._proc.on("exit", () => {
+      this.shutdown();
+    });
+    this._proc.on("SIGINT", () => {
+      this.shutdown();
+    });
+    this._adapterProcess.on("exit", () => {
+      this.shutdown();
+    });
+    this._adapterProcess.stdout.on("data", (data: Buffer) => {
+      this._handleData(data);
+    });
+    this._adapterProcess.stderr.on("data", (data: Buffer) => {
+      console.error(data.toString());
+      this.shutdown();
+    });
+  }
+
+  // callback to handle data written back by the adapter
+  _handleData(data: Buffer): void {
+    this._rawData = Buffer.concat([this._rawData, data]);
+    // the following code parses a message according to the protocol.
+    while (this._rawData.length > 0) {
+      // if we know what length we are expecting
+      if (this._contentLength >= 0) {
+        // we have enough data to check for the expected message
+        if (this._rawData.length >= this._contentLength) {
+          // first get the expected message
+          let message = this._rawData.toString("utf8", 0, this._contentLength);
+          // reduce the buffer by the message we got
+          this._rawData = this._rawData.slice(this._contentLength);
+          // reset the content length to ensure it is extracted for the next message
+          this._contentLength = -1;
+          // process the message
+          if (message.length > 0) {
+            try {
+              let msg = JSON.parse(message);
+              if (msg.type === "event") {
+                this._processEvent(msg);
+              } else if (msg.type === "response") {
+                this._processResponse(msg);
+              }
+            } catch (e) {
+              console.log("Invalid message");
+            }
+          }
+          continue; // there may be more complete messages to process
+        }
+      } else {
+        // if we don't know the length to expect, we need to extract it first
+        let idx = this._rawData.indexOf(TWO_CRLF);
+        if (idx !== -1) {
+          let header = this._rawData.toString("utf8", 0, idx);
+          let lines = header.split("\r\n");
+          for (let i = 0; i < lines.length; i++) {
+            let pair = lines[i].split(/: +/);
+            if (pair[0] === "Content-Length") {
+              this._contentLength = +pair[1];
+            }
+          }
+          this._rawData = this._rawData.slice(idx + TWO_CRLF.length);
+          continue;
+        }
+        // if we don't find the length we fall through and break
+      }
+      break;
+    }
+
+    //ask the user for the next command
+    this._reader.question("(dbg) ", (input: string) => {
+      this._dispatch(input);
+    });
+  }
+
+  _processEvent(event: DebugProtocol.Event) {
+    // to be implemented
```

Comment:
Looks like you're not quite done with this.

Reply:
This and _processEvent will be implemented in the next PR to not make this one too big. For now, they just confirm a message is received and parsed by printing it.